go/vt/vtctl/workflow/utils_test.go

```diff
@@ -0,0 +1,168 @@
+package workflow
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"os/exec"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"vitess.io/vitess/go/testfiles"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/etcd2topo"
+	"vitess.io/vitess/go/vt/topo/memorytopo"
+	"vitess.io/vitess/go/vt/topotools"
+)
+
+// TestUpdateKeyspaceRoutingRule confirms that the keyspace routing rules are updated correctly.
+func TestUpdateKeyspaceRoutingRule(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ts := memorytopo.NewServer(ctx, "zone1")
+	defer ts.Close()
+	routes := make(map[string]string)
+	for _, tabletType := range tabletTypeSuffixes {
+		routes["from"+tabletType] = "to"
+	}
+	err := updateKeyspaceRoutingRule(ctx, ts, "ks", "test", routes)
+	require.NoError(t, err)
+	rules, err := topotools.GetKeyspaceRoutingRules(ctx, ts)
+	require.NoError(t, err)
+	require.EqualValues(t, routes, rules)
+}
+
+// TestConcurrentKeyspaceRoutingRulesUpdates runs multiple keyspace routing rules updates concurrently to test
+// the locking mechanism.
+func TestConcurrentKeyspaceRoutingRulesUpdates(t *testing.T) {
+	ctx := context.Background()
+
+	ts := memorytopo.NewServer(ctx, "zone1")
+	defer ts.Close()
+	t.Run("memtopo", func(t *testing.T) {
+		testConcurrentKeyspaceRoutingRulesUpdates(t, ctx, ts)
+	})
+
+	etcdServerAddress := startEtcd(t)
+	log.Infof("Successfully started etcd server at %s", etcdServerAddress)
+	topoName := "etcd2_test" // "etcd2" is already registered on init(), so using a different name
+	topo.RegisterFactory(topoName, etcd2topo.Factory{})
+	ts, err := topo.OpenServer(topoName, etcdServerAddress, "/vitess")
+	require.NoError(t, err)
+	t.Run("etcd", func(t *testing.T) {
+		testConcurrentKeyspaceRoutingRulesUpdates(t, ctx, ts)
+		ts.Close()
+	})
+}
+
+func testConcurrentKeyspaceRoutingRulesUpdates(t *testing.T, ctx context.Context, ts *topo.Server) {
+	concurrency := 100
+	duration := 3 * time.Second
+
+	var wg sync.WaitGroup
+	wg.Add(concurrency)
+
+	stop := make(chan struct{})
```

Comment:
Simpler IMO to use a context.WithTimeout(duration) rather than the stop channel and the done channel used with time.After.